File: src/energy-converter/energy-converter.ts
import DistanceFormatter from './distance-formatter';
import VolumeFormatter from './volume-formatter';
import energyData, { EnergyEntry, EnergyType } from './data/energy';

export interface EnergyMatch {
  label: string;
  value: string;
}

class EnergyConverter {
  private readonly energyEntries: EnergyEntry[];
  private distanceFormatter: DistanceFormatter;
  private volumeFormatter: VolumeFormatter;

  constructor() {
    this.energyEntries = energyData;
    this.distanceFormatter = new DistanceFormatter();
    this.volumeFormatter = new VolumeFormatter();
  }

  convert(milliWattHour: number): EnergyMatch[] {
    const matches: EnergyMatch[] = [];
    const kiloWattHour = milliWattHour / 1_000_000;

    for (const entry of this.energyEntries) {
      const amount = kiloWattHour / entry.energy_capacity_kwh;
      const msg = this.format(entry, amount);
      matches.push({ label: entry.label, value: msg });
    }

    return matches;
  }

  private format(entry: EnergyEntry, value: number): string {
    switch (entry.type) {
      case EnergyType.VOLUME:
        return this.volumeFormatter.format(value);
      case EnergyType.DISTANCE:
        return this.distanceFormatter.format(value);
      default:
        throw new Error(`Unknown entry type: ${entry.type}`);
    }
  }
}

export default EnergyConverter;

File: cloud-run-function-kotlin/app/src/main/kotlin/org/phdezann/energy/service/EnergyConverter.kt
package org.phdezann.energy.service

import org.phdezann.energy.formater.DistanceFormatter
import org.phdezann.energy.formater.VolumeFormatter
import org.phdezann.energy.json.EnergyEntry
import org.phdezann.energy.json.EnergyEntryType
import org.phdezann.energy.model.EnergyMatch
import org.phdezann.energy.support.FileUtils

class EnergyConverter {

    private val distanceFormatter = DistanceFormatter()
    private val volumeFormatter = VolumeFormatter()
    private val energyEntries: Array<EnergyEntry>

    init {
        val json = FileUtils.read("energy.json")
        energyEntries = JsonSerializer().readValue(json, Array<EnergyEntry>::class.java)
    }

    fun convert(milliWattHour: Double): List<EnergyMatch> {
        val matches = mutableListOf<EnergyMatch>()
        val kiloWattHour = milliWattHour / 1_000_000

        for (entry in energyEntries) {
            val amount = kiloWattHour / entry.energyCapacityKwh
            val msg = format(entry, amount)
            matches.add(EnergyMatch(entry.label, msg))
        }

        return matches
    }

    private fun format(entry: EnergyEntry, value: Double): String {
        return when (entry.type) {
            EnergyEntryType.VOLUME -> volumeFormatter.format(value)
            EnergyEntryType.DISTANCE -> distanceFormatter.format(value)
        }
    }
}